// server/events.routes.js
import { Router } from "express";
import { db } from "./db.js";
import { requireAuth } from "./auth.routes.js";

const router = Router();
const h = (fn) => (req, res, next) => Promise.resolve(fn(req, res, next)).catch(next);

router.use(requireAuth);

router.get("/", h(async (req, res) => {
  const rows = await db.listEvents(req.userId, { from: req.query.from, to: req.query.to });
  res.json(rows.map(serialize));
}));

router.post("/", h(async (req, res) => {
  const { calKey, title, startsAt, endsAt, location, summary, allDay } = req.body ?? {};
  if (!title?.trim() || !startsAt || !endsAt)
    return res.status(400).json({ message: "제목과 시간은 필수입니다." });
  const cal = await db.findCalendar(req.userId, calKey ?? "personal");
  if (!cal) return res.status(400).json({ message: "존재하지 않는 캘린더입니다." });
  const row = await db.createEvent(req.userId, {
    calendar_id: cal.id, title: title.trim(),
    starts_at: startsAt, ends_at: endsAt,
    location: location ?? null, summary: summary ?? null, all_day: !!allDay,
  });
  res.json(serialize(row));
}));

router.patch("/:id", h(async (req, res) => {
  const map = { title: "title", startsAt: "starts_at", endsAt: "ends_at", location: "location", summary: "summary", allDay: "all_day" };
  const fields = {};
  for (const [k, col] of Object.entries(map)) {
    if (req.body?.[k] !== undefined) fields[col] = k === "allDay" ? !!req.body[k] : req.body[k];
  }
  if (!Object.keys(fields).length) return res.status(400).json({ message: "변경할 내용이 없습니다." });
  const row = await db.updateEvent(req.userId, req.params.id, fields);
  if (!row) return res.status(404).json({ message: "일정을 찾을 수 없습니다." });
  res.json(serialize(row));
}));

router.delete("/:id", h(async (req, res) => {
  const ok = await db.deleteEvent(req.userId, req.params.id);
  if (!ok) return res.status(404).json({ message: "일정을 찾을 수 없습니다." });
  res.status(204).end();
}));

function serialize(r) {
  return {
    id: r.id, calKey: r.cal_key, calColor: r.cal_color, title: r.title,
    startsAt: r.starts_at, endsAt: r.ends_at, allDay: r.all_day,
    location: r.location, summary: r.summary,
  };
}

export default router;
